python/cudf/cudf/io/parquet.py

```diff
@@ -790,9 +780,180 @@ def merge_parquet_filemetadata(filemetadata_list):
+        self.filename = None
+
+    def write_table(self, df):
+        """
+        Write a dataframe to the file/dataset
+        """
+        (
+            paths,
+            metadata_file_paths,
+            grouped_df,
+            offsets,
+            self.filename,
+        ) = _get_partitioned(
+            df,
+            self.path,
+            self.partition_cols,
+            preserve_index=self.common_args["index"],
+            filename=self.filename,
+        )
+
+        existing_cw_batch = defaultdict(dict)
+        new_cw_paths = []
+
+        def pairwise(iterable):
+            """
+            Generates a pair of `(it[i], it[i + 1] - it[i])` from iterable `it`
+            """
+            it = iter(iterable)
+            a = next(it, None)
+            for b in it:
+                yield (a, b - a)
+                a = b
+
+        for path, part_info, meta_path in zip(
+            paths, pairwise(offsets), metadata_file_paths
+        ):
+            if path in self.path_cw_map:  # path is a currently open file
+                cw_idx = self.path_cw_map[path]
+                existing_cw_batch[cw_idx][path] = part_info
+            else:  # path not currently handled by any chunked writer
+                new_cw_paths.append((path, part_info, meta_path))
+
+        # Write out the parts of grouped_df currently handled by existing cw's
+        for cw_idx, path_to_part_info_map in existing_cw_batch.items():
+            cw = self._chunked_writers[cw_idx][0]
+            # match found paths with this cw's paths and nullify partition info
+            # for partition_col values not in this batch
+            this_cw_part_info = [
+                path_to_part_info_map.get(path, (0, 0))
+                for path in self._chunked_writers[cw_idx][1]
+            ]
+            cw.write_table(grouped_df, this_cw_part_info)
+
+        # Create new cw for unhandled paths encountered in this write_table
+        new_paths, part_info, meta_paths = zip(*new_cw_paths)
+        self._chunked_writers.append(
+            (
+                ParquetWriter(new_paths, **self.common_args),
+                new_paths,
+                meta_paths,
+            )
+        )
+        new_cw_idx = len(self._chunked_writers) - 1
+        self.path_cw_map.update({k: new_cw_idx for k in new_paths})
+        self._chunked_writers[-1][0].write_table(grouped_df, part_info)
+
+    def close(self, return_metadata=False):
+        """
+        Close all open files and optionally return footer metadata as a binary
+        blob
+        """
+
+        metadata = [
+            cw.close(metadata_file_path=meta_path if return_metadata else None)
+            for cw, _, meta_path in self._chunked_writers
+        ]
+
+        if return_metadata:
+            return (
+                merge_parquet_filemetadata(metadata)
+                if len(metadata) > 1
+                else metadata[0]
+            )
+
+    def __del__(self):
+        self.close()
```

Comment:
I'll add the usual comment about `__del__` that it's not always guaranteed to be called, even at interpreter shutdown: https://docs.python.org/3/reference/datamodel.html#object.__del__

There are lots of warnings in SO not to use `__del__` as a destructor. The suggestion is to use `__enter__()`  and `__exit__()` methods instead, the latter of which will call `close()`.

This changes the API, but guarantees that the `close()` method will be called:

```python
with ParquetDatasetWriter(...) as pq_writer:
    pq_writer.write_table(df1)
    pq_writer.write_table(df2)
# pq_writer.close() will be called upon exiting the `with` statement
```

If we don't care that the `close()` method may not always be called, I think using `__del__` is OK, and I've seen it done in other parts of RAPIDS. If it's absolutely essential that all `ParquetDatasetWriter` objects be `close`'d, we might want to reconsider.

Reply:
> If it's absolutely essential that all `ParquetDatasetWriter` objects be `close`'d

It is.

Does cython's `__dealloc__` have the same limitation or is it more like the c++ destructor? I ask this because the non-partitioned `ParquetWriter` uses that.

Reply:
I've never been able to find a reliable reference on whether  `__dealloc__`  (which translates into the type's [tp_dealloc](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dealloc) method) has the same limitations. My understanding is that it does.

However, now that you mention it, I'm seeing that `ParquetWriter` is doing something explicitly unsafe, i.e., calling a Python method in its `__dealloc__` method. See [here](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dealloc) for why not to do that. The Cython docs here suggest just using `__del__` in those situations instead...

To be frank, I don't know about situations in which `__del__` or `__dealloc__` may not be called. We rely on `__dealloc__` within RAPIDS to free memory allocated in C++ and it has worked well. In any rare cases where it may not work, we wouldn't run into an error, "just" a memory leak. However, the `ParquetWriter` case may be different, where we need `__dealloc__`/`__del__` to work as expected for  correctness.

Reply:
The only user of `ParquetWriter` known to me is nvtabular and they wrap it in another `ThreadedWriter` class which does not have a destructor at all, but does have a [`close()`](https://github.dev/NVIDIA-Merlin/NVTabular/blob/90776815a13be5081af38c1b957252dd65c620ee/nvtabular/io/writer.py#L303). Their `close()` calls `ParquetWriter`'s `close()`.

So should we just remove the destructor altogether?
And then should we add `__enter__()` and `__exit__()`? How does that fare as python class design? Having two ways to use it.

Reply:
It's pretty standard. The `file` object in Python offers a similar interface, where you can explicitly close it with `close`, or use it within a `with` statement which will call `close` for you.

Calling `close` multiple times should be allowed, where subsequent calls to `close()` do nothing.